js/src/features/project/components/ProjectRoleForm.tsx

```diff
@@ -57,176 +70,196 @@ export const ProjectRoleForm: React.FC<Props> = ({
+    if (!!form[index].id) {
+      return indexOfEditableInput === index;
+    }
+    return true;
   }
 
-  function editPermission(index: number, newPermission: string) {
-    const role = project.roles[index];
+  function editInputRole(index: number) {
+    const field = form[index];
+    const value = inputRef.current!.state.value;
 
-    const permissions = role.permissions.includes(newPermission)
-      ? role.permissions.filter((permission) => permission !== newPermission)
-      : [...role.permissions, newPermission];
+    setIndexOfEditableField(index);
+    setIndexOfEditableInput(-1);
 
-    edit({
-      roles: {
-        ...role,
-        permissions,
+    setForm({
+      ...form,
+      [index]: {
+        ...field,
+        name: value,
       },
-      slug: project.slug,
     });
   }
 
-  function toggleCheckbox(index: number, permission: string) {
-    const role = project.roles[index];
+  function createRole(index: number, value: string) {
+    const field = form[index];
 
-    if (!role) {
-      createRole(index, permission);
-      return;
-    }
+    if (!!field.id) return;
 
-    editPermission(index, permission);
+    setForm({
+      ...form,
+      [index]: {
+        ...field,
+        name: value,
+      },
+    });
   }
 
-  function isChecked(value: string, index: number) {
-    return project.roles[index]?.permissions.some(
-      (permission) => permission === value,
-    );
-  }
+  function addFormField() {
+    const key = Object.keys(form).length;
 
-  function isDisabled(index: number) {
-    return !form.getFieldValue("roles")[index];
+    setForm({
+      ...form,
+      [key]: {
+        name: "",
+        permissions: [],
+      },
+    });
   }
 
-  function isOpenInput(index: number) {
-    if (!!project.roles[index]) {
-      return permissionIndexField === index;
-    }
-    return true;
-  }
+  useEffect(() => {
+    setForm(initialForm);
+    setIndexOfEditableInput(-1);
+    setIndexOfEditableField(-1);
+  }, [initialForm]);
 
   useEffect(() => {
-    form.setFieldsValue({ roles: project.roles });
-    setPermissionIndexField(-1);
-  }, [project.roles, form]);
+    if (indexOfEditableField !== -1 && indexOfEditableInput === -1) {
+      const field = form[indexOfEditableField];
+      const params = {
+        roles: field,
+        slug: project.slug,
+      };
+
+      if (!field.id && field.permissions.length > 0) {
+        create(params);
+      } else if (field.id) {
+        edit(params);
+      }
+
+      setIndexOfEditableInput(-1);
+      setIndexOfEditableField(-1);
+    }
+  }, [
+    form,
+    indexOfEditableField,
+    create,
+    edit,
+    project.slug,
+    indexOfEditableInput,
+  ]);
 
   return (
-    <Form
-      form={form}
-      onFinishFailed={onFinishFailed}
-      initialValues={initialValues}
-      className="role-form"
-    >
-      <Form.List name="roles">
-        {(fields, { add, remove }) => {
+    <ProjectRoleFormTableWrapper permissions={rolePermissions}>
```

Comment:
Maybe we can use here Ant Table component ?

Reply:
For this case it’s easier to make your table

Reply:
Agree with @Aliaksandrzaz. Ant Table requires dataSource. But we need only table grid.

Reply:
Now all information about a table stored in form. We can use it as dataSource.